lib/model/internal_link.dart

```diff
@@ -0,0 +1,253 @@
+// it says where to look in a list of messages, but it doesn't filter the list down.
+// In fact, from a brief look at server code, it seems to be *ignored*
+// if you include it in the `narrow` param in get-messages requests.
+// When you want to point the server to a location in a message list, you
+// you do so by passing the `anchor` param.
+Uri narrowLink(PerAccountStore store, Narrow narrow, {int? nearMessageId}) {
+  final apiNarrow = narrow.apiEncode();
+  final fragment = StringBuffer('narrow');
+  for (ApiNarrowElement element in apiNarrow) {
+    fragment.write('/');
+    if (element.negated) {
+      fragment.write('-');
+    }
+
+    if (element is ApiNarrowDm) {
+      final supportsOperatorDm = store.connection.zulipFeatureLevel! >= 177; // TODO(server-7)
+      element = element.resolve(legacy: !supportsOperatorDm);
+    }
+
+    fragment.write('${element.operator}/');
+
+    switch (element) {
+      case ApiNarrowStream():
+        final streamId = element.operand;
+        final name = store.streams[streamId]?.name ?? 'unknown';
+        final slugifiedName = _encodeHashComponent(name.replaceAll(' ', '-'));
+        fragment.write('$streamId-$slugifiedName');
+      case ApiNarrowTopic():
+        fragment.write(_encodeHashComponent(element.operand));
+      case ApiNarrowDmModern():
+        final suffix = element.operand.length >= 3 ? 'group' : 'dm';
+        fragment.write('${element.operand.join(',')}-$suffix');
+      case ApiNarrowPmWith():
+        final suffix = element.operand.length >= 3 ? 'group' : 'pm';
+        fragment.write('${element.operand.join(',')}-$suffix');
+      case ApiNarrowDm():
+        assert(false, 'ApiNarrowDm should have been resolved');
+      case ApiNarrowMessageId():
+        fragment.write(element.operand.toString());
+    }
+  }
+
+  if (nearMessageId != null) {
+    fragment.write('/near/$nearMessageId');
+  }
+
+  return store.account.realmUrl.replace(fragment: fragment.toString());
+}
+
+/// A [Narrow] from a given URL, on `store`'s realm.
+///
+/// Returns `null` if any of the operator/operand pairs are invalid.
+///
+/// Since narrow links can combine operators in ways our Narrow type can't
+/// represent, this can also return null for valid narrow links.
+///
+/// This can also return null for some valid narrow links that our Narrow
+/// type *could* accurately represent. We should try to understand these
+/// better, but some kinds will be rare, even unheard-of:
+///   #narrow/stream/1-announce/stream/1-announce (duplicated operator)
+///
+/// The passed `url` must appear to be a link to a Zulip narrow on the given
+/// `realm`.
+Narrow? parseInternalLink(Uri url, PerAccountStore store) {
+  if (!url.hasFragment) return null;
+  if (!url.hasScheme || url.host.isEmpty) return null;
+  if (!url.hasEmptyPath && (url.path != '/')) return null;
+
+  if (url.origin != store.account.realmUrl.origin) return null;
```

Comment:
My thinking of putting `hasFragment` up top was for optimization, since all internal links are defined in the fragment, it's a quick check if it exists or not, rather than a string comparision of the origin.

As for query, the mobile app rejects any urls with a query set, but the web UI ignores it. I'll reorder for the general-to-specific (I agree about the added clarity of that) and will follow the mobile behavior on rejecting queries.